File: go/pkg/parser/document_visitor.go
package parser

import "github.com/mojo-lang/lang/go/pkg/lang"

type DocumentVisitor struct {
	*BaseMojoParserVisitor
}

func NewDocumentVisitor() *DocumentVisitor {
	return &DocumentVisitor{}
}

func GetDocument(ctx IDocumentContext) *lang.Document {
	if ctx != nil {
		visitor := NewDocumentVisitor()
		return ctx.Accept(visitor).(*lang.Document)
	}
	return nil
}

func GetFollowingDocument(ctx IFollowingDocumentContext) *lang.Document {
	if ctx != nil {
		visitor := NewDocumentVisitor()
		return ctx.Accept(visitor).(*lang.Document)
	}
	return nil
}

func (d *DocumentVisitor) VisitDocument(ctx *DocumentContext) interface{} {
	lines := ctx.AllLineDocument()
	document := &lang.Document{}
	for _, line := range lines {
		lineDocument := &lang.LineDocument{}
		lineDocument.Line = line.GetText()
		//lineDocument.StartPosition = GetPosition(line.GetStart())
		//lineDocument.EndPosition = GetPosition(line.GetStop())

		document.Lines = append(document.Lines, lineDocument)
	}

	return document
}

func (d *DocumentVisitor) VisitFollowingDocument (ctx *FollowingDocumentContext) interface{} {
	document := &lang.Document{}
	return document
}

File: java/mojo-parser/src/main/java/org/mojolang/mojo/parser/DocumentVisitor.java
package org.mojolang.mojo.parser;

import com.google.protobuf.GeneratedMessageV3;
import org.antlr.v4.runtime.tree.TerminalNode;
import org.mojolang.mojo.lang.Document;
import org.mojolang.mojo.lang.LineDocument;

import java.util.List;

public class DocumentVisitor extends MojoBaseVisitor<com.google.protobuf.GeneratedMessageV3> {
    @Override
    public GeneratedMessageV3 visitDocument(MojoParser.DocumentContext ctx) {
        List<TerminalNode> lines = ctx.Line_document();
        Document.Builder builder = Document.newBuilder();

        for (TerminalNode line: lines) {
            LineDocument.Builder lineBuilder = LineDocument.newBuilder();
            lineBuilder.setLine(line.getText());
            lineBuilder.setStartPosition(Util.getPosition(line.getSymbol()));

            builder.addLines(lineBuilder);
        }

        return builder.build();
    }
}
